tests/Unit/Framework/Pypp.hpp

```diff
@@ -26,58 +26,40 @@
 #include "Utilities/TypeTraits/IsA.hpp"
 #include "Utilities/TypeTraits/IsStdArray.hpp"
 
-/// \ingroup TestingFrameworkGroup
-/// Contains all functions for calling python from C++
+/*!
+ * \ingroup TestingFrameworkGroup
+ * \brief Contains all functions for calling python from C++
+ */
 namespace pypp {
 namespace detail {
-template <typename R>
-R call_work(PyObject* python_module, PyObject* func, PyObject* args) {
-  PyObject* value = PyObject_CallObject(func, args);
-  Py_DECREF(args);  // NOLINT
-  if (value == nullptr) {
-    Py_DECREF(func);           // NOLINT
-    Py_DECREF(python_module);  // NOLINT
-    PyErr_Print();
-    throw std::runtime_error{"Function returned null"};
-  }
+template <typename T, typename ConversionClass>
+struct PackedTypeMatches : std::false_type {};
 
-  auto ret = from_py_object<R>(value);
-  Py_DECREF(value);  // NOLINT
-  return ret;
-}
+template <typename ConversionClass>
+struct PackedTypeMatches<typename ConversionClass::packed_container,
+                         ConversionClass> : std::true_type {};
 
-template <typename R, typename = std::nullptr_t>
-struct CallImpl {
-  template <typename... Args>
-  static R call(const std::string& module_name,
-                const std::string& function_name, const Args&... t) {
-    PyObject* python_module = PyImport_ImportModule(module_name.c_str());
-    if (python_module == nullptr) {
-      PyErr_Print();
-      throw std::runtime_error{std::string("Could not find python module.\n") +
-                               module_name};
-    }
-    PyObject* func =
-        PyObject_GetAttrString(python_module, function_name.c_str());
-    if (func == nullptr or not PyCallable_Check(func)) {
-      if (PyErr_Occurred()) {
-        PyErr_Print();
-      }
-      throw std::runtime_error{"Could not find python function in module.\n"};
-    }
-    PyObject* args = pypp::make_py_tuple(t...);
-    auto ret = call_work<R>(python_module, func, args);
-    Py_DECREF(func);           // NOLINT
-    Py_DECREF(python_module);  // NOLINT
-    return ret;
-  }
-};
-
-template <typename T, typename = std::nullptr_t>
+template <typename T, typename ConversionClassList, typename = std::nullptr_t>
 struct ContainerPackAndUnpack;
 
-template <>
-struct ContainerPackAndUnpack<DataVector, std::nullptr_t> {
+template <typename T, typename ConversionClassList>
+struct ContainerPackAndUnpack<
```

Comment:
What's the design reasoning for making this take an arbitrary conversion class list, rather than just asking new pack routines specialize the `ContainerPackAndUnpack` struct?

Reply:
ODR violations... There is some discussion above on this PR and a very nice comment by Will that explains that the issue with the specializations is that then every test file needs to define a unique specialization. However, I could reasonable see someone wanting to convert the same type (e.g. an EOS) but with different parameters in different files. E.g. I want a polytropic EOS with gamma=4/3 in file A and gamma=5/3 in file B. With different specializations in different files this leads to ODR violations. Even trying to fix this by marking some of the wrapping classes and functions as having internal linkage, this procedures remains brittle and errors will be impossible to debug for anyone except for the people who've had this conversation (currently 3 of us). Passing in the type list makes it so that you can specify any conversion in any local file and there will not be any weird C++ funny business. Seem alright?

Reply:
Aha, understood. I suppose ODR violations are still _technically_ possible here, though, right? If two different tests define a new class `struct EosPackAndUnpack` with different contents, and have identical conversion class lists (`tmpl::list<EosPackAndUnpack>`), then the same problem comes up, does it not?
(I'm not advocating for any further indirection, I'm just trying to come to a working grasp of how avoiding the ODR violation works)

Reply:
Yea, this is a bit tricky. The key thing is what @wthrowe said:

> For this PR, I think this means roughly that, if f has external linkage, then for any types T1...Tn, f(T1,...,Tn) must do the same thing in every translation unit in which it is called. (Similarly for template parameters.) I think the easiest way to do that is to encode any custom translations in a (possibly deduced) template parameter. Any customization a translation unit wants can be defined with internal linkage so there is no way any other translation unit could make a conflicting call.

This means that by making `pypp::call` explicitly depend on the locally defined (with internal linkage) struct we get a new definition of the function in the translation unit. Adding a specialization of the `ContainerPackAndUnpack` in a translation unit is not an explicit dependency of `pypp::call` and therefore you have one definition for all translation units. This is where the ODR violation comes in. It's straightforward to create a little example:
```
 $ cat test2.cpp    
#include <vector>
#include <memory>
#include <iostream>


void foo_a();
void foo_b();

int main() {
  foo_a();
  foo_b();
}

 $ cat test.hpp  
#pragma once

#include <iostream>

template <typename T>
struct Printer {
  void print() {
    std::cout << "Hello\n";
  }
};

template <typename T>
void printer() {
  T::print();
}

 $ cat fooa.cpp
#include <iostream>

#include "test.hpp"


namespace {
struct MyPrinter {
  static void print() {
    std::cout << "Hi from a\n";
  }
};
}  // namespace

template <>
struct Printer<double> {
  static void print() {
    std::cout << "Hello from a\n";
  }
};

void foo_a() {
  printer<MyPrinter>();
  printer<Printer<double>>();
}

 $ cat foob.cpp
#include <iostream>

#include "test.hpp"


namespace {
struct MyPrinter {
  static void print() {
    std::cout << "Hi from b\n";
  }
};
}  // namespace

template <>
struct Printer<double> {
  static void print() {
    std::cout << "Hello from b\n";
  }
};

void foo_b() {
  printer<MyPrinter>();
  printer<Printer<double>>();
}
```
The output is:
```
Hi from a
Hello from a
Hi from b
Hello from a
```